### PR TITLE
hwinfo: several improvements

### PR DESCRIPTION
**Pull request by bobvanderlinden**
## Description of changes

Follow-up from https://github.com/NixOS/nixpkgs/pull/334633#pullrequestreview-2244309876

This PR includes several improvements to the hwinfo package.
See the individual commits for more details.

The newly introduced tests pass.
The updater seems to be working.

This also fixes another possible use of `/usr/bin/`.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by brianmcgee — APPROVED** [review]
LGTM

File: pkgs/by-name/hw/hwinfo/package.nix [hunk 1]
```diff
@@ -63,6 +59,17 @@ stdenv.mkDerivation (finalAttrs: {
 
   passthru.tests = {
     version = testers.testVersion { package = finalAttrs.finalPackage; };
+    no-usr = testers.testEqualContents {
+      assertion = "There should be no /usr/ paths in the binaries";
+      # There is a bash script that refers to lshal, which is deprecated and not available in Nixpkgs.
+      # We'll allow this line, but nothing else.
+      expected = writeText "expected" ''
+        if [ -x /usr/bin/lshal ] ; then
```
Comment (bjornfor) [inline pkgs/by-name/hw/hwinfo/package.nix]: How does this work? It looks like an unfinished if-block?

Comment (bjornfor) [reply pkgs/by-name/hw/hwinfo/package.nix]: Ah, I understand now. The contents of the "expected" derivation output is compared to the contents of "actual".

File: pkgs/by-name/hw/hwinfo/package.nix [hunk 2]
```diff
@@ -30,9 +30,13 @@ stdenv.mkDerivation (finalAttrs: {
...
-    substituteInPlace Makefile --replace "/sbin" "/bin" --replace "/usr/" "/"
-    substituteInPlace src/isdn/cdb/Makefile --replace "lex isdn_cdb.lex" "flex isdn_cdb.lex"
-    substituteInPlace hwinfo.pc.in --replace "prefix=/usr" "prefix=$out"
+    substituteInPlace Makefile \
+      --replace-fail "/sbin" "/bin" \
+      --replace-fail "/usr/" "/"
+    substituteInPlace src/isdn/cdb/Makefile \
+      --replace-fail "lex isdn_cdb.lex" "flex isdn_cdb.lex"
+    substituteInPlace hwinfo.pc.in \
+      --replace-fail "prefix=/usr" "prefix=$out"
```
Comment (bjornfor) [inline pkgs/by-name/hw/hwinfo/package.nix:49]: This is not a pure formatting change, it also changes "--replace" into "--replace-fail".

**Review by bjornfor — APPROVED** [review]

File: pkgs/by-name/hw/hwinfo/package.nix (hunk 2 above)
Comment (bobvanderlinden) [reply pkgs/by-name/hw/hwinfo/package.nix:49]: Ah sorry, the commit message is indeed misleading 🤦 I'll reword it.

Comment (bobvanderlinden) [reply pkgs/by-name/hw/hwinfo/package.nix:49]: Reworded :+1: Thanks for the heads-up. I think the PR is ready for merge

**Pull request merged**